parquet-variant/src/main/java/org/apache/parquet/variant/Variant.java

```diff
@@ -0,0 +1,444 @@
+   * @param str the string to escape
+   * @return the escaped string
+   */
+  private static String escapeJson(String str) {
+    try (CharArrayWriter writer = new CharArrayWriter();
+        JsonGenerator gen = new JsonFactory().createGenerator(writer)) {
+      gen.writeString(str);
+      gen.flush();
+      return writer.toString();
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  /**
+   * Appends a quoted string to a StringBuilder. It is used when we know `str` doesn't contain any
+   * special characters that needs escaping. This is more performant than
+   * `sb.append(escapeJson(str))`.
+   * @param sb the StringBuilder to append to
+   * @param str the string to append
+   */
+  private static void appendQuoted(StringBuilder sb, String str) {
+    sb.append('"');
+    sb.append(str);
+    sb.append('"');
+  }
+
+  /** The format for a timestamp without time zone. */
+  private static final DateTimeFormatter TIMESTAMP_NTZ_FORMATTER = new DateTimeFormatterBuilder()
+      .append(DateTimeFormatter.ISO_LOCAL_DATE)
+      .appendLiteral('T')
+      .appendPattern("HH:mm:ss")
+      .appendFraction(MICRO_OF_SECOND, 6, 6, true)
+      .toFormatter(Locale.US);
+
+  /** The format for a timestamp with time zone. */
+  private static final DateTimeFormatter TIMESTAMP_FORMATTER = new DateTimeFormatterBuilder()
+      .append(TIMESTAMP_NTZ_FORMATTER)
+      .appendOffset("+HH:MM", "+00:00")
+      .toFormatter(Locale.US);
+
+  /** The format for a timestamp without time zone, truncating trailing microsecond zeros. */
+  private static final DateTimeFormatter TIMESTAMP_NTZ_TRUNC_FORMATTER = new DateTimeFormatterBuilder()
+      .append(DateTimeFormatter.ISO_LOCAL_DATE)
+      .appendLiteral('T')
+      .appendPattern("HH:mm:ss")
+      .optionalStart()
+      .appendFraction(MICRO_OF_SECOND, 0, 6, true)
+      .optionalEnd()
+      .toFormatter(Locale.US);
+
+  /** The format for a timestamp with time zone, truncating trailing microsecond zeros. */
+  private static final DateTimeFormatter TIMESTAMP_TRUNC_FORMATTER = new DateTimeFormatterBuilder()
+      .append(TIMESTAMP_NTZ_TRUNC_FORMATTER)
+      .appendOffset("+HH:MM", "+00:00")
+      .toFormatter(Locale.US);
+
+  private static Instant microsToInstant(long microsSinceEpoch) {
+    return Instant.EPOCH.plus(microsSinceEpoch, ChronoUnit.MICROS);
+  }
+
+  private static void toJsonImpl(
+      byte[] value, byte[] metadata, int pos, StringBuilder sb, ZoneId zoneId, boolean truncateTrailingZeros) {
+    switch (VariantUtil.getType(value, pos)) {
+      case OBJECT:
+        handleObject(value, pos, (size, idSize, offsetSize, idStart, offsetStart, dataStart) -> {
+          sb.append('{');
+          for (int i = 0; i < size; ++i) {
+            int id = readUnsigned(value, idStart + idSize * i, idSize);
+            int offset = readUnsigned(value, offsetStart + offsetSize * i, offsetSize);
```

Comment:
The logic here is copied in multiple places. I think it would be better to avoid copying. Instead, why not use an approach similar to `getFieldAtIndex` combined with `handleObject`? You could either use an `Iterator` or accept a lambda for each field.